File: src/round_key.rs
use core::mem;

// Constants for key expansion
const CONSTVAL: [u32; 8] = [0xC3EFE9DB, 0x44626B02, 0x79E27C8A, 0x78DF30EC, 0x715EA49E, 0xC785DA0A, 0xE04EF22A, 0xE5C40957];

pub	fn generate(key: [u32; 16] ) -> [u32; 144] {
    // Convert key to u32 array
    let key_ptr = key.as_ptr().cast::<[u32; 4]>();

    // Check if key is aligned
    let key_is_aligned = key_ptr.align_offset(mem::align_of::<[u32; 4]>()) == 0;

    // Read key
    let mut rk_t = if key_is_aligned {
        unsafe { key_ptr.read() }
    } else {
        unsafe { key_ptr.read_unaligned() }
    };

    // Initialize round key
    let mut rk : [u32; 144] = [0; 144];

    // Copy key to round key
    for i in 0..4 {
        rk[i] = rk_t[i];
    }

    // Generate round key
    for i in 0..24 {
        let t0 = CONSTVAL[i % 4].rotate_left(i as u32);
        let t1 = t0.rotate_left(1);
        let t2 = t1.rotate_left(1);
        let t3 = t2.rotate_left(1);
        rk_t[0] = rk_t[0].wrapping_add(t0).rotate_left(1);
        rk_t[1] = rk_t[1].wrapping_add(t1).rotate_left(3);
        rk_t[2] = rk_t[2].wrapping_add(t2).rotate_left(6);
        rk_t[3] = rk_t[3].wrapping_add(t3).rotate_left(11);

        rk[6 * i + 0] = rk_t[0];
        rk[6 * i + 1] = rk_t[1];
        rk[6 * i + 2] = rk_t[2];
        rk[6 * i + 3] = rk_t[1];
        rk[6 * i + 4] = rk_t[3];
        rk[6 * i + 5] = rk_t[1];
    }

    rk
}

File: encryption/round_key.go
package encryption

import "lea/bitops"

var constval = [8]uint32{0xC3EFE9DB, 0x44626B02, 0x79E27C8A, 0x78DF30EC, 0x715EA49E, 0xC785DA0A, 0xE04EF22A, 0xE5C40957}

func Generate(key [16]uint32) [144]uint32 {
	if len(key) != 16 {
		panic("Key must be an array of 16 uint32s")
	}

	// Initialize round key array
	rk := [144]uint32{}

	// Copy key to round key
	for i := 0; i < 4; i++ {
		rk[i] = key[i]
	}

	// Generate round key
	var rkT = make([]uint32, 4)
	copy(rkT, key[:4]) // Copy first four elements to temporary array

	for i := 0; i < 24; i++ {
		t0 := bitops.RotateLeft32(constval[i%4], uint(i))
		t1 := bitops.RotateLeft32(t0, 1)
		t2 := bitops.RotateLeft32(t1, 1)
		t3 := bitops.RotateLeft32(t2, 1)

		rkT[0] = bitops.RotateLeft32(rkT[0]+t0, 1)
		rkT[1] = bitops.RotateLeft32(rkT[1]+t1, 3)
		rkT[2] = bitops.RotateLeft32(rkT[2]+t2, 6)
		rkT[3] = bitops.RotateLeft32(rkT[3]+t3, 11)

		j := 6 * i
		rk[j+0] = rkT[0]
		rk[j+1] = rkT[1]
		rk[j+2] = rkT[2]
		rk[j+3] = rkT[1] // Assuming the repeated use of rkT[1] is intentional
		rk[j+4] = rkT[3]
		rk[j+5] = rkT[1] // Repeated again
	}

	return rk
}
